clippy_lints/src/disallowed_types.rs

```diff
@@ -66,6 +68,11 @@ impl DisallowedTypes {
     }
 
     fn check_res_emit(&self, cx: &LateContext<'_>, res: &Res, span: Span) {
+        if in_external_macro(cx.sess(), span) {
```

Comment:
`in_external_macro` is kinda... buggy? I assume, it doesn't detects macros with spanned quote for some reason.

So, if you have a macro defined like this:

```rust
#[proc_macro_attribute]
pub fn use_std_hash_map(_args: TokenStream, input: TokenStream) -> TokenStream {
    let mut item = syn::parse_macro_input!(input as syn::ItemFn);
    let span = item.block.brace_token.span;
    let block = item.block;
    item.block = syn::parse_quote_spanned! {
        span =>
        {
            let _ = std::collections::HashMap::<i32, i32>::default();
        }
    };
    quote! {#item}.into()
}
```

It still triggers this lint, I think you'll need a `clippy_utils::is_from_proc_macro` check additionally.

Reply:
Added (except for poly trait ref case - `PolyTraitRef` doesn't implement required trait there).